rig-core/src/providers/gemini/completion.rs

```diff
@@ -0,0 +1,649 @@
+            generation_config.max_output_tokens = Some(max_tokens);
+        }
+
+        let request = GenerateContentRequest {
+            contents: full_history
+                .into_iter()
+                .map(|msg| Content {
+                    parts: vec![Part { text: msg.content }],
+                    role: match msg.role.as_str() {
+                        "system" => Some("model".to_string()),
+                        "user" => Some("user".to_string()),
+                        "assistant" => Some("model".to_string()),
+                        _ => None,
+                    },
+                })
+                .collect(),
+            generation_config: Some(generation_config),
+            safety_settings: None,
+            tools: Some(
+                completion_request
+                    .tools
+                    .into_iter()
+                    .map(Tool::from)
+                    .collect(),
+            ),
+            tool_config: None,
+            system_instruction: Some(Content {
+                parts: vec![Part {
+                    text: "system".to_string(),
+                }],
+                role: Some("system".to_string()),
+            }),
+        };
+
+        tracing::info!("Request: {:?}", request);
+
+        let response = self
+            .client
+            .post(&format!("/v1beta/models/{}:generateContent", self.model))
+            .json(&request)
+            .send()
+            .await?
+            .error_for_status()?
+            .json::<ApiResponse<GenerateContentResponse>>()
+            .await?;
+
+        match response {
+            ApiResponse::Ok(response) => Ok(response.try_into()?),
+            ApiResponse::Err(err) => Err(CompletionError::ResponseError(err.message)),
+        }
+    }
+}
+
+impl From<completion::ToolDefinition> for Tool {
+    fn from(tool: completion::ToolDefinition) -> Self {
+        Self {
+            function_declaration: FunctionDeclaration {
+                name: tool.name,
+                description: tool.description,
+                parameters: None, // tool.parameters, TODO: Map Gemini
+            },
+            code_execution: None,
+        }
+    }
+}
+
+impl TryFrom<GenerateContentResponse> for completion::CompletionResponse<GenerateContentResponse> {
+    type Error = CompletionError;
+
+    fn try_from(response: GenerateContentResponse) -> Result<Self, Self::Error> {
+        match response.candidates.as_slice() {
+            [ContentCandidate { content, .. }, ..] => Ok(completion::CompletionResponse {
+                choice: completion::ModelChoice::Message(
+                    content.parts.first().unwrap().text.clone(),
+                ),
+                raw_response: response,
+            }),
+            _ => Err(CompletionError::ResponseError(
+                "No candidates found in response".into(),
+            )),
+        }
+    }
+}
+
+impl TryFrom<serde_json::Value> for GenerationConfig {
```

Comment:
You may not need this `TryFrom` implementation. Instead of doing `GenerationConfig::try_from(completion_request.additional_params.unwrap_or_default())?;`, you can do `serde_json::from_value::<GenerationConfig>(completion_request.additional_params.unwrap_or_default())?`. Since `GenerationConfig` has Option for each field, if the additional params don't contain the generation_config field, it won't fail. 

This won't work though if field X of `completion_request.additional_params` is a string but field X of GenerationConfig expects a float for example.

Reply:
@mateobelanger is there a reason why you didn't change this?